api/envoy/admin/v2alpha/config_dump.proto

```diff
@@ -0,0 +1,118 @@
+import "envoy/api/v2/rds.proto";
+import "envoy/config/bootstrap/v2/bootstrap.proto";
+
+import "google/protobuf/any.proto";
+
+import "gogoproto/gogo.proto";
+
+// [#protodoc-title: ConfigDump]
+
+// The /config_dump admin endpoint uses this wrapper message to maintain and serve arbitrary
+// configuration information from any component in Envoy.
+// TODO(jsedgwick) In the future, we may want to formalize this further with an RPC for config_dump,
+// and perhaps even with an RPC per config type. That strategy across all endpoints will allow for
+// more flexibility w.r.t. protocol, serialization, parameters, etc.
+message ConfigDump {
+  // This map is serialized and dumped in its entirety at the /config_dump endpoint.
+  //
+  // Keys should be a short descriptor of the config object they map to. For example, Envoy's HTTP
+  // routing subsystem might use "routes" as the key for its config, for which it uses the message
+  // RouteConfigDump as defined below. In the future, the key will also be used to filter the output
+  // of the /config_dump endpoint.
+  map<string, google.protobuf.Any> configs = 1 [(gogoproto.nullable) = false];
+}
+
+// This message describes the bootstrap configuration that Envoy was started with. This includes
+// any CLI overrides that were merged.
+message BootstrapConfigDump {
+  envoy.config.bootstrap.v2.Bootstrap bootstrap = 1 [(gogoproto.nullable) = false];
+}
+
+// Envoy's listener manager fills this message with all currently known listeners.
+message ListenersConfigDump {
+  // This is the top level, "transactional," version information in the last processed LDS
+  // discovery response. If there are only static bootstrap listeners, this field will be "".
+  string version_info = 1;
+
+  // Describes a dynamically loaded cluster via the LDS API.
+  message DynamicListener {
+    // This is the top level, "transactional," version information in the last processed LDS
+    // discovery response. If there are only static bootstrap listeners, this field will be "".
+    string version_info = 1;
+
+    // The listener config.
+    envoy.api.v2.Listener listener = 2;
+  }
+
+  // The statically loaded listener configs.
+  repeated envoy.api.v2.Listener static_listeners = 2 [(gogoproto.nullable) = false];
+
+  // The dynamically loaded active listeners These are listeners that are available to service
+  // data plane traffic.
+  repeated DynamicListener dynamic_active_listeners = 3 [(gogoproto.nullable) = false];
+
+  // The dynamically loaded warming listeners These are listeners that are currently undergoing
+  // warming in preparation to service data plane traffic.
+  repeated DynamicListener dynamic_warming_listeners = 4 [(gogoproto.nullable) = false];
+
+  // The dynamically loaded draining listeners These are listeners that are currently undergoing
+  // draining in preparation to stop servicing data plane traffic.
+  repeated DynamicListener dynamic_draining_listeners = 5 [(gogoproto.nullable) = false];
```

Comment:
I'm wondering how this all ties together with the end objective of being able to dump and then relaunch an Envoy with thee same configuration? It's hard to bring back into the precise warming/draining state, I don't think you want to either. It would be good to document how each field is to be interpreted, e.g. "throw away the draining listeners when recreating config, take the warming and active and static, merge them together and then deliver in static config or LDS", etc.

Reply:
Sure, I can add some more comments. I do think it's important to dump warming/draining/etc. since it gives a much more complete view of what is going on. I do also think that we should add config_dump query params that allow filtering out certain types of data which might make reload easier.

Reply:
@htuch "being able to dump and then relaunch an Envoy with thee same configuration" - are we expecting people build tools around this to this work? or what is the exact usecase? Sorry I do not have context of this PR hence this question

Reply:
@ramaraochavali yes, ultimately, we think it would be pretty awesome if there was a tool that could take the output of config_dump and produce a relatively close approximation of the config using only static resources. Would be great for debugging.